alphadia/workflow/optimization.py

```diff
@@ -0,0 +1,684 @@
+        """See base class. The update rule is
+            1) calculate the deviation of the predicted mz values from the observed mz values,
+            2) take the mean of the endpoints of the central 99% of these deviations, and
+            3) multiply this value by 1.1.
+        This is implemented by the ci method for the estimator.
+
+        Returns
+        -------
+        float
+            The proposed new value for the search parameter.
+
+        """
+        return 1.1 * self.workflow.calibration_manager.get_estimator(
+            self.estimator_group_name, self.estimator_name
+        ).ci(df, 0.99)
+
+    def _get_feature_value(
+        self, precursors_df: pd.DataFrame, fragments_df: pd.DataFrame
+    ):
+        return len(precursors_df)
+
+
+class AutomaticMS2Optimizer(AutomaticOptimizer):
+    def __init__(
+        self,
+        initial_parameter: float,
+        workflow,
+        reporter: None | reporting.Pipeline | reporting.Backend = None,
+    ):
+        """See base class. This class automatically optimizes the MS2 tolerance parameter by tracking the number of precursor identifications and stopping when further changes do not increase this number."""
+        self.parameter_name = "ms2_error"
+        self.estimator_group_name = "fragment"
+        self.estimator_name = "mz"
+        self.feature_name = "precursor_count"
+        super().__init__(initial_parameter, workflow, reporter)
+
+    def _check_convergence(self, current_step: int = -1):
+        """Optimization should stop if continued narrowing of the MS2 parameter is not improving the number of precursor identifications.
+        This function checks if the previous rounds of optimization have led to a meaningful improvement in the number of identifications.
+        If so, it continues optimization and appends the proposed new parameter to the list of parameters. If not, it stops optimization and sets the optimal parameter attribute.
+
+        Notes
+        -----
+            Because the check for an increase in identifications requires two previous rounds, the function will also initialize for another round of optimization if there have been fewer than 3 rounds.
+
+        Parameters
+        ----------
+        current_step: int
+            The current step in the optimization process. By default it is set to -1, which prevents the optimizer from converging unless min_steps has been set to 0.
+
+        Returns
+        -------
+        bool
+            True if the convergence conditions are met, False otherwise.
+
+        """
+
+        return (
+            len(self.history_df) > 2
+            and self.history_df[self.feature_name].iloc[-1]
+            < 1.1 * self.history_df[self.feature_name].iloc[-2]
+            and self.history_df[self.feature_name].iloc[-1]
+            < 1.1 * self.history_df[self.feature_name].iloc[-3]
+            and current_step >= self.workflow.config["calibration"]["min_steps"] - 1
+        )
```

Comment:
try to move this logic up in the class hierarchy (duplicated 5 times)

Reply:
I haven't done this because there is a good chance these methods will become parameter-specific as we improve the automatic calibration. I could change it to neaten up the code, but there is a good chance we would have to change them back again.

Reply:
if this chance is above 80% then it's fine .. otherwise I would always opt for implementing according to current requirements, because _cloudy the future is_ ;-)